extensions/gitpod-remote/src/ports.ts

```diff
@@ -0,0 +1,122 @@
+/*---------------------------------------------------------------------------------------------
+ *  Copyright (c) Gitpod. All rights reserved.
+ *--------------------------------------------------------------------------------------------*/
+import * as vscode from 'vscode';
+import * as grpc from '@grpc/grpc-js';
+import * as util from 'util';
+import { GitpodExtensionContext, GitpodWorkspacePort } from 'gitpod-shared';
+import { PortsStatus, PortsStatusRequest, PortsStatusResponse } from '@gitpod/supervisor-api-grpc/lib/status_pb';
+import { RetryAutoExposeRequest, TunnelVisiblity } from '@gitpod/supervisor-api-grpc/lib/port_pb';
+
+export async function getSupervisorPorts(context: GitpodExtensionContext) {
+	let supervisorPortList: PortsStatus.AsObject[] = [];
+	try {
+		supervisorPortList = await new Promise<PortsStatus.AsObject[]>((resolve, reject) => {
+			const req = new PortsStatusRequest();
+			const evts = context.supervisor.status.portsStatus(req, context.supervisor.metadata);
+			evts.on('error', reject);
+			evts.on('data', (resp: PortsStatusResponse) => resolve(resp.getPortsList().map(p => p.toObject())));
+		});
+	} catch (e) {
+		context.logger.error('Could not fetch ports info from supervisor', e);
+	}
+	return supervisorPortList;
+}
+
+export function observePortsStatus(context: GitpodExtensionContext): [vscode.EventEmitter<PortsStatus.AsObject[]>, { dispose: () => any }] {
+	const onPortUpdate = new vscode.EventEmitter<PortsStatus.AsObject[]>();
+	let run = true;
+	let stopUpdates: Function | undefined;
+	(async () => {
+		while (run) {
+			try {
+				const req = new PortsStatusRequest();
+				req.setObserve(true);
+				const evts = context.supervisor.status.portsStatus(req, context.supervisor.metadata);
+				stopUpdates = evts.cancel.bind(evts);
```

Comment:
What we use `stopUpdates` for? `cancelContext`?

Reply:
I just copied this xD, but it will be used in the dispose method to cancel the portsStatus request